### DISCUSSION TITLE
Exploring InterModule Communication

### DISCUSSION THREAD

**Discussion by kasif** [oqtane/oqtane.framework#4592] [Q&A] [answered]
I am Working on a Project where I have created multiple Modules.
All the modules are configured and in running state independently.
Now I have requirement where I need to set up communication between the modules.
Like I want to access Module A's properties, services, repositories from Module B and manipulate the data.

What are the different ways to achieve it, and which one is the best?

Thanks,
Mohd Kasif Nawaz.

**Comment by sbwalker** [comment, marked as answer]
You mentioned "inter-module communication" which usually refers to passing state between different components at runtime. However I think what you are actually referring to is how to leverage repositories, services, etc... from different modules during development (which is something completely different).

The easiest way to accomplish your development goal is to create an Oqtane project which contains multiple modules:

```
Module.sln
/Client
  Module.Client.csproj
  /Modules
    /Module1
      Index.razor (make sure namespaces are unique for each module)
    /Module2
      Index.razor
  /Services
    Module1Service.cs
    Module2Service.cs
/Server
  Module.Server.csproj
  /Controllers
    Module1Controller.cs
    Module2Controller.cs
  /Repository
    Module1Repository.cs
    Module2Repository.cs
/Shared
  Module.Shared.csproj
  /Models
    Module1Model.cs
    Module2Model.cs
```

Note that your solution will be compiled into 3 assemblies (which internally contain both Module1 and Module2 implementations) - and these need to be deployed to the Oqtane Framework /bin folder. Using this approach you can easily reference Module1's services, etc... from Module2 and vice versa.

If you look closely at the folder organization above you will notice that it matches the Oqtane Framework's folder organization - the Oqtane Framework is essentially a project which contains many modules.

**Reply by kasif** [reply]
Yeah, thanks for understanding my query and providing the solution.
However, considering me a beginner to oqtane, I am unable to understanding the provided solution.
See, below is the screenshot of the Oqtane Project with the modules associated with it.
So are you referring making changes in main Oqtane Project, or in specific Module Itself?
![image](https://github.com/user-attachments/assets/4b03f99b-d97e-4dad-9ad9-98f00fbfaa50)
If The changes are to be made in Module, then do we need to make changes in all the modules?
Below is snapshot of a Module that contains client, server, shared project.
![image](https://github.com/user-attachments/assets/5a1c3b0c-45c9-4917-adb7-60349fb89425)

Please explain me at a beginner level.
Thanks,
Mohd Kasif

**Comment by sbwalker** [comment]
@kasif based on your screenshot it looks like you have created 4 different modules. It also looks like you renamed the Oqtane Framework folder to "oqtane.framework-5.1.2" which will cause difficulty with the module "Package" projects debug.cmd file which does post-build deployment of DLLs to the appropriate locations based on folder path.

My earlier post was explaining the a single module solution can contain multiple modules. So basically you can take the code from your 4 modules and merge them into a single module solution using the folder organization I outlined above.

**Reply by kasif** [reply]
I have not renamed it. I have just downloaded 5.1.2 version of oqtane, and its coming by this name from the start.
furthermore, so currently, I should create a new Module that contains merged Modules(all the 4) and organize them as you have shown.
But can't the functionality be achieved by having 4 different modules?

**Reply by sbwalker** [reply]
Yes, you could use 4 separate modules - however you would need to create the necessary dependencies between them. For example if ModuleA needs to call a Service from ModuleB then you will need to include the necessary project reference between the module projects so that ModuleA is aware of ModuleB. These dependencies can become challenging to manage as sometimes you will end up with circular dependencies - so it is much easier to aggregate them together.

**Reply by kasif** [reply]
ok thanks,
I will take the recommended approach.
